components/supervisor/pkg/terminal/terminal.go

```diff
@@ -146,35 +150,49 @@ func (m *Mux) doClose(alias string, gracePeriod time.Duration) error {
 	return nil
 }
 
-func (term *Term) gracefullyShutdownProcess(gracePeriod time.Duration) error {
+func (term *Term) gracefullyShutdownProcess(sig syscall.Signal) (err error) {
 	if term.Command.Process == nil {
-		// process is alrady gone
+		// process is already gone
 		return nil
 	}
-	if gracePeriod == 0 {
+
+	if sig == unix.SIGKILL {
 		return term.shutdownProcessImmediately()
 	}
 
-	err := term.Command.Process.Signal(unix.SIGTERM)
-	if err != nil {
-		return err
-	}
-	schan := make(chan error, 1)
-	go func() {
-		_, err := term.Wait()
-		schan <- err
-	}()
-	select {
-	case err = <-schan:
+	// 1. ask to terminate
+	pid := term.Command.Process.Pid
+	// only SIGHUP is propagated to child processes by shell
+	// see https://www.gnu.org/software/bash/manual/html_node/Signals.html
+	if sig != unix.SIGHUP {
+		// otherwise we try to send signal to the foreground process group
+		// it is aligned with how other supervisor systems, i.e. systemd or Docker
+		// see https://veithen.io/2014/11/16/sigterm-propagation.html
+		var pgrp int
+		pgrp, err = unix.IoctlGetInt(term.fd, unix.TIOCGPGRP)
 		if err == nil {
-			// process is gone now - we're good
-			return nil
+			pid = pgrp
+		} else {
+			log.WithError(err).Error("terminal: failed to resolve the foreground process group")
 		}
-		log.WithError(err).Warn("unexpected terminal error")
-	case <-time.After(gracePeriod):
+	}
+	alive := syscall.Kill(pid, sig) == nil
+	if !alive {
+		return
 	}
 
-	// process did not exit in time. Let's kill.
+	// 2. wait to terminate
+	for i := time.Duration(0); alive && i < term.closeGracePeiod; i += time.Second {
```

Comment:
I think we can get rid of the `closeGracePeriod`, because there is an outer graceperiod that is imposed by kubernetes that will kill supervisor eventually.